src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+
+        # Combine pulse layout into a single dataframe.
+        # TODO: samples_per_pulses is currently assumed to be constant
+        pulse_layout = pd.DataFrame({'count': num_pulses,
+                                     'first': pulse_first,
+                                     'last': pulse_last,
+                                     'length': samples_per_pulse})
+
+        return pulse_ids, pulse_layout
+
+    def _reshape_pulses_by_train(self, data, num_pulses, samples_per_pulse,
+                                 first_pulse_offset=None):
+        """Reshape traces by train into traces by train and pulses."""
+
+        pulses_start = first_pulse_offset or self._first_pulse_offset
+        pulses_end = pulses_start + num_pulses * samples_per_pulse
+
+        if pulses_end > data.shape[-1]:
+            raise ValueError(f'trace axis too short for {num_pulses} pulses '
+                             f'located at {pulses_start}:{pulses_end}]')
+
+        return data[..., pulses_start:pulses_end].reshape(
+            *data.shape[:-1], num_pulses, samples_per_pulse)
+
+    def _reshape_flat_pulses(self, data, out, pulse_ids, samples_per_pulse,
+                             first_pulse_offset=None):
+        """Reshape traces by train into traces by pulses."""
+
+        return _reshape_flat_pulses(
+            data, out, pulse_ids, samples_per_pulse,
+            first_pulse_offset or self._first_pulse_offset, self._clock_ratio)
+
+    def _build_train_pulse_coords(self, out, pulse_dim, pulse_layout):
+        """Build xarray coordinates for pulses by train."""
+
+        coords = {'trainId': self._raw_key.train_id_coordinates()}
+
+        # TODO: Create this from pulse index
+        if pulse_dim == 'pulseId':
+            coords['pulseId'] = self._pulses.peek_pulse_ids()
+        elif pulse_dim == 'pulseIndex':
+            coords['pulseIndex'] = np.arange(out.shape[1], dtype=np.int32)
+        elif pulse_dim == 'pulseTime':
+            coords['pulseTime'] = np.arange(
+                out.shape[1], dtype=np.int32) * pulse_layout['length'].max()
+        else:
+            raise ValueError(f'invalid pulse dimension `{pulse_dim}`')
+
+        return coords
+
+    def _build_sample_coords(self, data, sample_dim):
+        """Build xarray coordinates for samples."""
+
+        samples = np.arange(data.shape[-1], dtype=np.int32)
+
+        if sample_dim == 'time':
+            return {'time': samples * self.sampling_period}
+        elif sample_dim == 'sample':
+            return {'sample': samples}
+        else:
+            raise ValueError('sample_dim must be `time` or `sample`')
+
+    def _shape_edge_result(self, edges, amplitudes, index_or_coords,
+                           shape, squeeze_edges, with_amplitudes):
+        """Shape output data of edge finding methods."""
+
+        if shape == 'edges':
+            from . import DelayLineDetector
+
+            edges = DelayLineDetector._build_reduced_pd(
```

Comment:
Obviously this should move to a generic module, I simply imported it directly for time's sake.

@takluyver  Your comment on splitting up the method really payed off btw.